src/Raven.Server/Documents/Indexes/Static/V8/JavaScriptReduceOperation.cs

```diff
@@ -0,0 +1,206 @@
+using Raven.Server.Documents.Indexes.MapReduce;
+using Raven.Server.Documents.Patch.V8;
+using Raven.Server.ServerWide;
+using Sparrow;
+using Sparrow.Json;
+using Sparrow.Server;
+using V8.Net;
+using Raven.Server.Extensions.V8;
+using Raven.Server.Documents.Indexes.Static.Utils;
+using Raven.Client.ServerWide.JavaScript;
+using Raven.Server.Documents.Patch;
+
+namespace Raven.Server.Documents.Indexes.Static.V8
+{
+    public class JavaScriptReduceOperation : JavaScriptReduceOperationBase
+    {
+
+        public JavaScriptIndexUtils JavaScriptIndexUtils { get; }
+        public JavaScriptUtilsV8 JavaScriptUtilsV8 { get; }
+        public V8Engine Engine { get; }
+        public V8EngineEx EngineEx { get; }
+
+        public InternalHandle Reduce { get; }
+        public InternalHandle Key { get; }
+
+        public JavaScriptReduceOperation(JavaScriptIndexUtils javaScriptIndexUtils, ScriptFunctionInstance keyJint, Engine engineJint,
+            InternalHandle reduce, InternalHandle key)
+        : base(JavaScriptEngineType.V8, keyJint, engineJint)
+        {
+            if (reduce.IsUndefined || reduce.IsNull)
+                throw new ArgumentNullException(nameof(reduce));
+            InternalHandle reduceAux = reduce;
+            Reduce = new InternalHandle(ref reduceAux, true);
+
+            if (key.IsUndefined || key.IsNull)
+                throw new ArgumentNullException(nameof(key));
+            InternalHandle keyAux = key;
+            Key = new InternalHandle(ref keyAux, true);
+
+            JavaScriptIndexUtils = javaScriptIndexUtils;
+            JavaScriptUtilsV8 = (JavaScriptUtilsV8)JavaScriptIndexUtils.JavaScriptUtils;
+
+            EngineEx = (V8EngineEx)JavaScriptIndexUtils.EngineHandle;
+            Engine = EngineEx;
+        }
+
+        ~JavaScriptReduceOperation()
+        {
+            Reduce.Dispose();
+            Key.Dispose();
+        }
+        
+        public override IEnumerable IndexingFunction(IEnumerable<dynamic> items)
+        {
+            try
+            {
+                EnsureGroupItemCreated();
+                foreach (DynamicBlittableJson item in items)
+                {
+                    if (_groupedItems.TryGetValue(item.BlittableJson, out var list) == false)
+                    {
+                        list = new List<BlittableJsonReaderObject>();
+                        _groupedItems[item.BlittableJson] = list;
+                    }
+                    list.Add(item.BlittableJson);
+                }
+                foreach (var item in _groupedItems.Values)
+                {
+                    Engine.ResetCallStack();
+                    Engine.ResetConstraints();
+
+                    InternalHandle jsRes = InternalHandle.Empty;
+                    try
+                    {
+                        using (var jsGrouping = ConstructGrouping(item))
+                        {
+#if DEBUG
+                            Engine.MakeSnapshot("reduce");
```

Comment:
what is this for?

Reply:
for JS entities (InternalHandle) leakage detection (is performed for map and reduce separately in debug version)